hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java

```diff
@@ -175,6 +186,142 @@ public String getPollingTime() {
                 .getConfigurationValue(TenantConfigurationKey.POLLING_TIME_INTERVAL, String.class).getValue());
     }
 
+    /**
+     * Returns the configured minimum polling interval.
+     *
+     * @return current {@link TenantConfigurationKey#MIN_POLLING_TIME_INTERVAL}.
+     */
+    @Override
+    public String getMinPollingTime() {
+        return systemSecurityContext.runAsSystem(() -> tenantConfigurationManagement
+                .getConfigurationValue(TenantConfigurationKey.MIN_POLLING_TIME_INTERVAL, String.class).getValue());
+    }
+
+    /**
+     * Returns the count to be used for reducing polling interval while calling
+     * {@link ControllerManagement#getPollingTimeForAction()}.
+     *
+     * @return configured value of
+     *         {@link TenantConfigurationKey#MAINTENANCE_WINDOW_POLL_COUNT}.
+     */
+    @Override
+    public int getMaintenanceWindowPollCount() {
+        return systemSecurityContext.runAsSystem(() -> tenantConfigurationManagement
+                .getConfigurationValue(TenantConfigurationKey.MAINTENANCE_WINDOW_POLL_COUNT, Integer.class).getValue());
+    }
+
+    /**
+     * Returns polling time based on the maintenance window for an action.
+     * Server will reduce the polling interval as the start time for maintenance
+     * window approaches, so that at least these many attempts are made between
+     * current polling until start of maintenance window. Poll time keeps
+     * reducing with MinPollingTime as lower limit
+     * {@link TenantConfigurationKey#MIN_POLLING_TIME_INTERVAL}. After the start
+     * of maintenance window, it resets to default
+     * {@link TenantConfigurationKey#POLLING_TIME_INTERVAL}.
+     *
+     * @param action
+     *            id the {@link Action} for which polling time is calculated
+     *            based on it having maintenance window or not
+     *
+     * @return current {@link TenantConfigurationKey#POLLING_TIME_INTERVAL}.
+     */
+    @Override
+    public String getPollingTimeForAction(final Action action) {
+        if (action == null || !action.hasMaintenanceSchedule() || action.isMaintenanceScheduleLapsed()) {
+            return getPollingTime();
+        }
+
+        JpaAction jpaAction = getActionAndThrowExceptionIfNotFound(action.getId());
+        return (new EventTimer(getPollingTime(), getMinPollingTime(), ChronoUnit.SECONDS))
+                .timeToNextEvent(getMaintenanceWindowPollCount(), jpaAction.getMaintenanceWindowStartTime().get());
+    }
+
+    /**
+     * EventTimer to handle reduction of polling interval based on maintenance
+     * window start time. Class models the next polling time as an event to be
+     * raised and time to next polling as a timer. The event, in this case the
+     * polling, should happen when timer expires. Class makes use of java.time
+     * package to manipulate and calculate timer duration.
+     */
+    private class EventTimer {
+
+        String defaultEventInterval;
```

Comment:
fields can be private and final because only accessed and set within constructor, rigth?